### PR TITLE
Some example models

### PR DESCRIPTION
**Pull request by dirmeier**
Hello Nick,
I've created some example models. Maybe they are interesting/useful to you.

Best,
Simon

### CODE REVIEW

**Comment by dirmeier** [comment]
The Wishart-prior for the covariance-matrices does not make too much sense, though. I guess an inverse one is better: https://arxiv.org/pdf/1408.4050.pdf ?

**Comment by goldingn** [comment]
Nice one, thanks so much for all your effort on these!

There are a couple of things I would suggest changing to make all the examples simpler and easier for newcomers to navigate. I'll put them in a code review now

File: vignettes/example_models.Rmd [hunk 1]
```diff
@@ -33,14 +33,65 @@ Below are a few examples of common statistical models implemented in greta.
 
 <hr>
 
+```{r child='examples/linear_uninformative.Rmd'}
```
Comment (goldingn) [inline vignettes/example_models.Rmd]: It's great to have this demo of the different priors you can use in greta. I think they don't quite fit in the 'common models' section though.

Could you put them in a new section, with a title like 'common Bayesian priors', perhaps just showing the options for priors, like:
```r
 and a little discussion abou

File: vignettes/example_models.Rmd [hunk 2]
```diff
@@ -61,6 +112,19 @@ The following sections provide greta implementations of some of these example mo
 
 <hr>
 
+## Stan models
```
Comment (goldingn) [inline vignettes/example_models.Rmd]: This is awesome!

File: vignettes/example_models.Rmd [hunk 3]
```diff
@@ -33,14 +33,65 @@ Below are a few examples of common statistical models implemented in greta.
...
+```
+
+<hr>
+
 ```{r child='examples/hierarchical_linear.Rmd'}
 ```
 
+<hr>
+
+```{r child='examples/hierarchical_linear_slopes.Rmd'}
```
Comment (goldingn) [inline vignettes/example_models.Rmd]: I think this should go in the common models section

File: vignettes/example_models.Rmd [hunk 4]
```diff
@@ -33,14 +33,65 @@ Below are a few examples of common statistical models implemented in greta.
...
+```
+
+<hr>
+
+```{r child='examples/hierarchical_linear_general.Rmd'}
```
Comment (goldingn) [inline vignettes/example_models.Rmd]: I think these three are a bit more advanced, and should probably go in a separate section (Bayesian neural nets aren't likely to be common models for more first-time greta users!)

File: vignettes/examples/bayesian_neural_network.Rmd [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+neural_network <- function (x)
+{
+  # this can be arbitrarily complex, e.g. multiple hidden layers
+  x %*% weights
+}
+  
+N <- 100
+p <- 10
+
+set.seed(23)
```
Comment (goldingn) [inline vignettes/examples/bayesian_neural_network.Rmd]: could the data generation please go in a separate block above?

File: vignettes/examples/hierarchical_linear_general.Rmd [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+A hierarchical, Bayesian linear regression model using the iris data, with random intercepts and slopes for each of the three species.
+The slopes and intercepts for each species are *correlated* in this example. We allow every species to have a species specific slope for `Sepal.Length`.
+
+
+```{r hierarchical_linear_general_greta}
+int  <- normal(0, 10)
+coef <- normal(0, 10)
+sd   <- cauchy(0, 3, truncation = c(0, Inf))
+
+n.species  <- length(unique(iris$Species))
```
Comment (goldingn) [inline vignettes/examples/hierarchical_linear_general.Rmd]: `n.species` -> `n_species` for consistency

File: vignettes/examples/hierarchical_linear_general.Rmd [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+species_id <- as.numeric(iris$Species)
+
+Z     <- matrix(0, nrow(iris), n.species * 2)
+gamma <- zeros(n.species * 2)
+
+for (s in unique(species_id)) {
+  Z[species_id == s, s * 2  - 1] <- 1
```
Comment (goldingn) [inline vignettes/examples/hierarchical_linear_general.Rmd]: I think this is a simpler (and probably easier for users) way of creating the same matrix:
```r
Z <- model.matrix(~ Species + Sepal.Length * Species - 1, data = iris)
```

It's in a slightly different order (intercepts first), but otherwise the same

File: vignettes/examples/hierarchical_linear_general.Rmd [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  Z[species_id == s, s * 2]  <- iris$Sepal.Length[species_id == s]
+  ranef_sd <- diag(2)
+  gamma[(s * 2 - 1):(s * 2)] <- multivariate_normal(rep(0, 2), ranef_sd)
```
Comment (goldingn) [inline vignettes/examples/hierarchical_linear_general.Rmd]: as it's currently set up, this is the same (the prior has the same density) as:
```r
gamma <- normal(0, 1, dim = n.species * 2)
```
That might be simpler for people to understand, if that's what you wanted to go for. The coefficients will still be correlated (in the posterior), though it you wanted to explicitly model the correlations it would require a variable for the variance-covariance matrix `ranef_sd`

File: vignettes/examples/hierarchical_linear_slopes.Rmd [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+```{r hierarchical_linear_slopes_greta}
+# linear model parameters
+int <- normal(0, 10)
+coef <- normal(0, 10)
+sd <- cauchy(0, 3, truncation = c(0, Inf))
+
+species_id <- as.numeric(iris$Species)
+
+# random intercepts
+species_random_intercept_sd <- lognormal(0, 1)
```
Comment (goldingn) [inline vignettes/examples/hierarchical_linear_slopes.Rmd]: I wonder of these object names could be shortened a bit? I like that they are explicit, but it's a bit unwieldy when they are all combined on a single line.

File: vignettes/example_models.Rmd (hunk 1 above)
Comment (goldingn) [reply vignettes/example_models.Rmd]: I just remembered I wrote the following a while ago. Might be worth adding them at some point too:

```r
# this is smoothed-out approximation to the true spike-and-slab prior that
# replaces the discrete indicator variable with a continuous variable.
# 'spikiness' sets the amount of spike on this, values closer to 0 should be
# closer to the discrete version, but may take longer to converge. Values
# greater than 1 will break everything. Don't do that.
spike_slab <- function (spikiness = 0.1, slabiness = 10, dim = NULL) {
  stopifnot(spikiness < 1)
  slab <- normal(0, slabiness, dim = dim)
  spike <- gamma(spikiness, spikiness, dim = dim)
  spike * slab
}

horseshoe <- function (tau = 1, dim = NULL) {
  lambda <- cauchy(0, 1, truncation = c(0, Inf), dim = dim)
  sd <- tau ^ 2 * lambda ^ 2
  normal(zeros_like(sd), sd, dim = dim)
}

ponyshoe <- function (tau = 1, eta = 3, s = 1, dim = NULL) {
  tau2 <- tau ^ 2
  c2 <- inverse_gamma(alpha = eta / 2, beta = eta * s ^ 2)
  lambda <- cauchy(0, 1, truncation = c(0, Inf), dim = dim)
  lambda2 <- lambda ^ 2
  lambda_tilde <- (c2 * lambda2) / (c2 * tau2 * lambda2)
  sd <- tau2 * lambda_tilde ^ 2
  normal(zeros_like(sd), sd, dim = dim)
}
```

File: vignettes/examples/linear_lasso.Rmd [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+A simple, one-variable Bayesian linear regression model using an exponential-normal prior. 
+A exponential-normal prior can be interpreted like an equivalent to frequentist LASSO.
+
+```{r linear_lasso_greta}
+# variables & priors
+int  <- variable()
+sd   <- inverse_gamma(1, 1)
+
+lambda <- gamma(1, 1)
+tau  <- exponential(0.5 * lambda**2)
```
Comment (goldingn) [inline vignettes/examples/linear_lasso.Rmd]: I'm surprised `a**b` works already in greta, I never explicitly coded it up! Might be better to use `^` for consistency and familiarity for new users. Also spaces for the style guide, so:
```r
tau  <- exponential(0.5 * lambda ^ 2)
```

File: vignettes/examples/lightspeed.Rmd [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+       27, 31, 27, 26, 33, 26, 32, 32, 24, 39, 28, 24, 25, 32, 25, 
+       29, 27, 28, 29, 16, 23)
+n <- length(y)
+```
+</div>
+
+#### greta code
+```{r lightspeed_greta}
+beta  <- variable()
+sigma <- variable(lower=0)
```
Comment (goldingn) [inline vignettes/examples/lightspeed.Rmd]: I'm trying to follow the [tidyverse style guide](http://style.tidyverse.org/) where possible in the greta codebase and examples, since it's becoming the de facto standard.

Could you please put [spaces](http://style.tidyverse.org/syntax.html#spacing) around the `=`

File: vignettes/examples/multiple_linear_lasso.Rmd [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+A multi-variable Bayesian linear regression model using an exponential-normal prior for the coefficients.
+
+```{r multiple_linear_lasso_greta}
+# the predictors as a matrix
+design <- as.matrix(attitude[, 2:7])
+
+int   <- normal(0, 10)
+sd <- cauchy(0, 3, truncation = c(0, Inf))
+
+tau   <- zeros(ncol(design))
```
Comment (goldingn) [inline vignettes/examples/multiple_linear_lasso.Rmd]: This is the more greta-y (and slightly more efficient) way to define these:
```r
tau <- exponential(0.5, dim = ncol(design)) 
coefs <- normal(0, tau)
```

File: vignettes/examples/multiple_linear_multilogit.Rmd [hunk 13]
```diff
@@ -0,0 +1,31 @@
+### Multiple multinomial regression
+
+A multi-variable Bayesian linear regression model using the iris data.
+
+```{r multiple_linear_multilogit_greta}
+data(iris)
+
+library(tensorflow)
+op <- greta::.internals$nodes$constructors$op
```
Comment (goldingn) [inline vignettes/examples/multiple_linear_multilogit.Rmd]: there's `imultilogit()` now, which takes one fewer columns that the number of response columns, since one is the baseline/constrast

File: vignettes/examples/multiple_linear_multilogit.Rmd [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+# Only choose a small subset until categprical takes matrix arguments
+idxs <- c(1:10, 51:60, 101:111)
+n <- length(idxs)
+
+X <- as_data(cbind(1, iris[idxs,1:4]))
+y <- model.matrix(~Species, iris[idxs,])
+P <- ncol(X)
+K <- ncol(y)
+
+beta <-normal(0, 5, dim=c(P, K))
```
Comment (goldingn) [inline vignettes/examples/multiple_linear_multilogit.Rmd]: so this would be:
```r
beta <- normal(0, 5, dim = c(P, K - 1))
eta <- X %*% beta
prob <- imultilogit(eta)
```

File: vignettes/examples/multiple_linear_warpbreaks.Rmd [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+```{r hierarchical_linear_warpbreaks_greta}
+data("warpbreaks")
+head(warpbreaks)
+
+X <- as_data(model.matrix(breaks ~ wool + tension, warpbreaks))
+y <- as_data(warpbreaks$breaks)
+
+beta <- zeros(ncol(X))
+beta[1]  <- variable()
+for (i in seq(2, ncol(X)))
```
Comment (goldingn) [inline vignettes/examples/multiple_linear_warpbreaks.Rmd]: This would be a little tidier and more efficient:

```r
int <- variable()
coefs <- normal(0, 5, dim = ncol(X) - 1)
beta <- c(int, coefs)
```

File: vignettes/examples/poisson.Rmd [hunk 16]
```diff
@@ -31,6 +31,6 @@ beta <- normal(0, 10, dim = n_env)
 
 # model
 linear_predictor <- alpha + X %*% beta
-lambda <- log(linear_predictor)
+lambda <- exp(linear_predictor)
```
Comment (goldingn) [inline vignettes/examples/poisson.Rmd]: :+1:

File: vignettes/get_started.R [hunk 17]
```diff
@@ -0,0 +1,59 @@
+## ----setup, include = FALSE----------------------------------------------
```
Comment (goldingn) [inline vignettes/get_started.R]: looks like this was auto-generated by knitr and accidentally committed, could you please delete it and recommit?

**Comment by goldingn** [comment]
moving this to be against dev, since I'm not planning another release until those dev changes are all merged in

**Comment by goldingn** [comment]
>The Wishart-prior for the covariance-matrices does not make too much sense, though. I guess an inverse one is better: arxiv.org/pdf/1408.4050.pdf ?

I didn't see any Wishart's in these examples, but yeah, inverse Wishart is normally what's used. There isn't a separate distribution for that since it's just:
```r
covar <- solve(wishart(df, Sigma))
```

**Comment by dirmeier** [comment]
Cool, I'll implement your suggestions and add the *spike and slab* and the other ones...

File: vignettes/examples/hierarchical_linear_general.Rmd (hunk 7 above)
Comment (dirmeier) [reply vignettes/examples/hierarchical_linear_general.Rmd]: Thanks, I don't know why I didn't do it like this right from the start ...

**Comment by dirmeier** [comment]
> I didn't see any Wishart's in these examples, but yeah, inverse Wishart is normally what's used. 

I removed the (inverse) Wishart part, since using it as a prior on the covariance seems to cause an error. You can see that below, for instance. If I set `gamma_sd <- solve(wishart(10, diag(2)))`, Tensorflow throws an `ValueError`. If I use a fixed covariance like `diag(2)` this doesn't happen.
Is that due to model misspecification or a Tensorflow issue?

```R
int  <- normal(0, 10)
coef <- normal(0, 10)
sd   <- cauchy(0, 3, truncation = c(0, Inf))

n_species  <- length(unique(iris$Species))
species_id <- as.numeric(iris$Species)

Z <- model.matrix(~ Species + Sepal.Length * Species , data = iris)

gamma <- zeros(n_species * 2)
for (s in unique(species_id)) {
  # does throw a warning and error
  gamma_sd <- solve(wishart(10, diag(2)))
  # this would work: gamma_sd <- diag(2)
  gamma[c(s, s + n_species)] <- multivariate_normal(rep(0, 2), gamma_sd) 
}

wi <- as_data(iris$Sepal.Width)
Z  <- as_data(Z)

mu <- int + coef * wi + Z %*% gamma

distribution(iris$Sepal.Length) <- normal(mu, sd)
m <- model(gamma)
d <- mcmc(m)
```

File: vignettes/examples/hierarchical_linear_general.Rmd (hunk 8 above)
Comment (dirmeier) [reply vignettes/examples/hierarchical_linear_general.Rmd]: Yeah, I am aware. I had an inverse Wishart for the covariance, but that throws an TF error (see below). However, I find this formulation still more valuable, since it shows explicitely how the gammas, should be constructed. Once I figure out, why the Wishart does not work here, I'd have changed it back.

**Comment by dirmeier** [comment]
Hi Nick, I think I worked on all your reviews. Travis is however failing now, maybe because the dev branch uses tensorflow 1.8 and tensorflow-probability? I can't seem to fix this, so should just close this PR and open a new one on the dev branch?

**Comment by dirmeier** [comment]
@goldingn The issue was with `opt` not working anymore with the newer tensorflow/tf-probability. Exchanging it with `mcmc` in the Bayesian neural network example fixed it. I'll look into why `opt` throws an error here, though.

File: vignettes/get_started.R (hunk 17 above)
Comment (goldingn) [reply vignettes/get_started.R]: Sorry, this request wasn't clear. Could you please delete (and remove from git tracking) the *.R files that were generated by knitr? They aren't used by the package but clutter up the directory.

**Comment by goldingn** [comment]
Awesome!

I'm going to refactor the optimiser code on the dev branch soon anyway, so we can switch it back once that's done.

I recommended deleting the *.R files, but I'll just merge this now and do that myself.

Thanks so much for these!
